### PR TITLE
Add random seed during initialization

### PR DESCRIPTION
**Pull request by maxjiang93**
It's important to set all distributed workers to the same manual seed, else different workers will initialize to a different set of weights for the parameters and the model will not converge afterwards.

### CODE REVIEW

**Comment by sparticlesteve** [comment]
The broadcast in DDP takes care of that. Or, at least, I believe it should.

**Comment by maxjiang93** [comment]
No it doesn't... If we don't seed it manually to the same value it will give different initializations for weights and biases, and the model does not converge at all :p I  was wrestling with this bug for a week and finally found this to be the issue. If you check the DDP example in pytorch it also does manual seeding for the same reason.
[](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html)

**Comment by sparticlesteve** [comment]
Initialization will be different, but there's this broadcast in the DDP constructor right here:
https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L298

There is no seed set on this page: https://pytorch.org/docs/master/notes/ddp.html
and it describes the broadcast:

> Construction: The DDP constructor takes a reference to the local module, and broadcasts state_dict() from the process with rank 0 to all other processes in the group to make sure that all model replicas start from the exact same state

**Comment by sparticlesteve** [comment]
If you have an example showing the behavior, I'd be interested to see it.

**Comment by sparticlesteve** [comment]
I wrote a quick little test to confirm:

```python
    # ResNet50
    model = torchvision.models.resnet50().to(device)
    print('Initial weight sum:', sum([w.sum() for w in model.state_dict().values()]).item())

    # Wrap model for distributed training
    model = torch.nn.parallel.DistributedDataParallel(
        model, device_ids=[local_rank], output_device=local_rank)
    print('After broadcast:', sum([w.sum() for w in model.state_dict().values()]).item())
```

```
7: Initial weight sum: 53584.26953125
5: Initial weight sum: 53170.11328125
2: Initial weight sum: 53281.82421875
6: Initial weight sum: 53320.94140625
0: Initial weight sum: 52922.49609375
4: Initial weight sum: 53407.07421875
1: Initial weight sum: 53171.84375
3: Initial weight sum: 53428.0859375
3: After broadcast: 52922.49609375
4: After broadcast: 52922.49609375
5: After broadcast: 52922.49609375
2: After broadcast: 52922.49609375
6: After broadcast: 52922.49609375
7: After broadcast: 52922.49609375
1: After broadcast: 52922.49609375
0: After broadcast: 52922.49609375
```

Maybe you can try something similar in your code to see if this was really your issue.

**Comment by maxjiang93** [comment]
That's interesting. For my model, I initially did the opposite thing. I manually set a different random seed for each node (so that they produce a different shuffle for dataloading) and the model did not converge. The only thing I changed afterwards was to set them to the same seed (inspired by the example demo page) and a slightly different sampling scheme and it converged beautifully. I can look more into it to investigate what exactly broke the previous code and let you know!

**Comment by sparticlesteve** [comment]
Ok, yeah let me know.

And anyway thanks for trying to fix potential issues. These benchmark models are only run on synthetic data so they are not meant to converge, but the code serves as a useful example so I do want things to be correct.

**Comment by maxjiang93** [comment]
Hey Steve,
 
I went back and checked again and it turns out what made the difference was actually the data sampler rather than the random seeding, so I guess we don't really need this change. You are right about the broadcasts. Sorry about the confusion!

Max

**Comment by sparticlesteve** [comment]
Ok thanks Max.
